Improve handlers and state docs (#1289)
software-mansion/react-native-gesture-handler: docs/docs/state.md

sidebar_label: Handler State
---

As described in ["About Gesture Handlers"](about-handlers.md), gesture handlers can be treated as ["state machines"](https://en.wikipedia.org/wiki/Finite-state_machine).
At any given time, each handler instance has an assigned state that can change when new touch events occur or can be forced to change by the touch system in certain circumstances.

> A gesture handler can have six possible states:

Reviewer: ```suggestion
A gesture handler can be in one of the six possible states:
```

There are 6 states overall, but the handler can only be in one (a state machine)

Author: excellent

## State flows

The most typical flow of state is when a gesture handler picks up on an initial touch event then recognizes it then acknowledges its ending then resets itself back to the initial state. 

The flow looks as follows (longer arrows represent that there are possibly more touch events received before the state changes):

[`UNDETERMINED`](#undetermined) -> [`BEGAN`](#began) ------> [`ACTIVE`](#active) ------> [`END`](#end) -> [`UNDETERMINED`](#undetermined)

Another possible flow is when a handler receives touches that cause a recognition failure:

[`UNDETERMINED`](#undetermined) -> [`BEGAN`](#began) ------> [`FAILED`](#failed) -> [`UNDETERMINED`](#undetermined)

> Another possible flow is when a handler *does* properly recognize the gesture but then is interrupted by the touch system causing it to cancel. In that case the flow looks as follows:

Reviewer: ```suggestion
At last, when a handler does properly recognize the gesture but then is interrupted by the touch system. In that case, the gesture recognition is canceled and the flow looks as follows:
```

Removed repetition and (I hope) kinda improved wording.

Author: well done

[`UNDETERMINED`](#undetermined) -> [`BEGAN`](#began) ------> [`ACTIVE`](#active) ------> [`CANCELLED`](#cancelled) -> [`UNDETERMINED`](#undetermined)

## States

The section below describes all possible handler states:

### UNDETERMINED

> This is the initial state of each handler. It can be found before gestures have occurred and after they've completed. 

Reviewer: ```suggestion
This is the initial state of each handler and it goes into this state after it's done with recognizing a gesture.
```

For me, "before gestures have occurred and after they've completed" isn't precise enough, we want to signify the end of recognizing a gesture.

Author: just a quick change to -> "after it's done recognizing a gesture"

### FAILED

> A handler has received some touches but for some reason won't recognize them. Its state will **not** become `ACTIVE` but instead become `FAILED`. Afterwards, it is reset to `UNDETERMINED`. An example of this is if a finger travels more distance than a defined `maxDist` property allows.

Reviewer: ```suggestion
A handler has received some touches but for some reason didn't recognize them. An example of this is if a finger travels more distance than a defined `maxDist` property allows. In that situation, the handler won't become active and will fail instead. Afterward, it's state will be reset from to `UNDETERMINED`. 
```

"Afterward, it is reset to `UNDETERMINED`" – I think this kinda drifts away from a state machine abstraction.

I'm not sure about this suggestion, it surely can be worded better.

Author: I will make a small change to this, please review
